### PR TITLE
Stream::merge does not end prematurely if one stream is delayed

### PR DESCRIPTION
**Pull request by matklad**
r? @yoshuawuyts 

The bug was that if `left` returned `Poll::Pending`, the right could have returnrf `Poll::Ready(None)`, and we would consider the stream as finished, although we haven't really polled the left part thoroughly enough. 

I also don't understand why we need `cx.waker().wake_by_ref();` here: we are returning `Poll::Ready`, and calling `wake_by_ref` is only required if you return `Pending`?

### CODE REVIEW

**Comment by matklad** [comment]
Question (unrelated to PR mostly): is it a contract of `Stream` trait that, after returning `Poll::Ready(None)` once, all subsequent calls to `poll_next` should return `Poll::Ready(None)` as well? I think both initial and this version of `merge` assume this behavior, but it is not documented. I think merge doesn't neccaary have to require this contract from substreams, if we change it's internal state to 

```rust
enum MergeState {
    Both(L, R),
    Left(L),
    Right(R),
}
```

File: tests/stream.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    let t1 = task::spawn(async move {
+        let mut xs = Vec::new();
+        while let Some(x) = s.next().await {
+            xs.push(x);
+        }
+        xs
+    });
+
+    task::block_on(async move {
+        task::sleep(std::time::Duration::from_millis(5)).await;
```
Comment (stjepang) [inline tests/stream.rs]: Does this sleep have any significance? Perhaps it'd be safer to pick a larger duration, maybe 500ms?

Comment (matklad) [reply tests/stream.rs]: without this sleep, the test can pass by accident, as we would not poll a `Poll::Pending` out of the channel. Basically, we need to ensure that the `t` tasks is polled before we do a send via channel. Bumped the dealay to 500ms: we can trivially dial it back if we find test-suite to be too long!

Comment (matklad) [reply tests/stream.rs]: to be more precise, the unfortunate timing here can't lead to false positive: if the code is correct, the test will always pass. However, timing can lead to a false-negative, where a buggy code would pass the test. 

I've considered making this deterministic by writing a stupid `Stream` which answers with `Pending` the first time it is polled and with `Ready(Some(92))` the second time. In the end, I've decided to go with channels + sleep to simulate a similar effect: writing a stream impl manually can easily introduce errors of it's own.

**Comment by matklad** [comment]
Hm, judging by the `StreamExt::fuse` method, we don't actually have this guarantee. so we should call `.fuse` ourselves int the `Merge` ctor

**Comment by yoshuawuyts** [comment]
> Hm, judging by the `StreamExt::fuse` method, we don't actually have this guarantee. so we should call `.fuse` ourselves int the `Merge` ctor

Took some time to think about this, and you're entirely right. Phew, you just caught a whole lot of bugs here. Thanks so much for this patch!

**Comment by matklad** [comment]
@yoshuawuyts i’ve even found one (debatable though) in std::iter::Chain: https://github.com/rust-lang/rust/issues/66031 as a result of yesterday’s thinking about cancellation. Still, haven’t figured our cancellation itself :D

**Comment by matklad** [comment]
added fusing as well

File: Cargo.toml [hunk 2]
```diff
@@ -55,3 +55,11 @@ futures-preview = { version = "=0.3.0-alpha.19", features = ["async-await"] }
...
 futures-channel-preview = "=0.3.0-alpha.19"
 futures-util-preview = "=0.3.0-alpha.19"
+
+[[test]]
+name = "stream"
+required-features = ["unstable"]
+
+[[example]]
+name = "tcp-ipv4-and-6-echo"
+required-features = ["unstable"]
```
Comment (yoshuawuyts) [inline Cargo.toml]: Oh, I wasn't even aware this was a thing. TIL, thank you!

**Pull request merged**